[coro] Lower `llvm.coro.await.suspend.handle` to resume with tail call

zmodem opened the pull request:
The C++ standard requires that symmetric transfer from one coroutine to another is performed via a tail call. Failure to do so is a miscompile and often breaks programs by quickly overflowing the stack.

Until now, the coro split pass tried to ensure this in the `addMustTailToCoroResumes()` function by searching for `llvm.coro.resume` calls to lower as tail calls if the conditions were right: the right function arguments, attributes, calling convention etc., and if a `ret void` was sure to be reached after traversal with some ad-hoc constant folding following the call.

This was brittle, as the kind of implicit variants required for a tail call to happen could easily be broken by other passes (e.g. if some instruction got in between the `resume` and `ret`), see for example 9d1cb18d19862fc0627e4a56e1e491a498e84c71 and 284da049f5feb62b40f5abc41dda7895e3d81d72.

Also the logic seemed backwards: instead of searching for possible tail call candidates and doing them if the circumstances are right, it seems better to start with the intention of making the tail calls we need, and forcing the circumstances to be right.

Now that we have the `llvm.coro.await.suspend.handle` intrinsic (since f78688134026686288a8d310b493d9327753a022) which corresponds exactly to symmetric transfer, change the lowering of that to also include the `resume` part, always lowered as a tail call.

zmodem committed:
stuff

zmodem committed:
format

mtrofin commented:
Does this address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043? Could you add a note there in that direction - a few folks were looking at going at the direction @jyknight was suggesting there, and it'd be good to have closure on the topic.

zmodem commented:
> Does this address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043? Could you add a note there in that direction - a few folks were looking at going at the direction @jyknight was suggesting there, and it'd be good to have closure on the topic.

Thanks for the pointer! If I understand things correctly:

* Yes, my patch addresses the problem of instructions in the suspend block getting in the way of lowering symmetric transfer as a tail call.
* With my patch, something could still put instructions between the llvm.coro.await.suspend.handle and the suspend, and those would become unreachable. I don't think that's a problem in practice though. The frontend shouldn't do that, and I don't think any transform would move instructions across the intrinsic call.
* My patch doesn't address the more general concern in that thread about that edge being a strange representation.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -1523,24 +1442,16 @@ struct SwitchCoroutineSplitter {
-    // Adding musttail call to support symmetric transfer.
-    // Skip targets which don't support tail call.
-    //
-    // FIXME: Could we support symmetric transfer effectively without musttail
```
Maybe we need to remain the FIXME

zmodem replied:
Done.

ChuanqiXu9 reviewed (commented):
The higher level idea looks fine.

A detail is that, in this patch, we must not mark `llvm.coro.await.suspend.handle` as nounwind. Previously, `llvm.coro.await.suspend.handle` may be marked as nounwind if the `await_suspend` is noexcept. But we can't do it in this patch since `llvm.coro.await.suspend.handle` will include `resume` calls, which technically is not nounwind.

zmodem committed:
reinstate fixme

zmodem committed:
llvm.coro.await.suspend.handle calls must not be nounwind

zmodem commented:
> The higher level idea looks fine.

Thanks!

> A detail is that, in this patch, we must not mark `llvm.coro.await.suspend.handle` as nounwind. Previously, `llvm.coro.await.suspend.handle` may be marked as nounwind if the `await_suspend` is noexcept. But we can't do it in this patch since `llvm.coro.await.suspend.handle` will include `resume` calls, which technically is not nounwind.

Done.

mtrofin commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -1056,6 +1083,25 @@ void CoroCloner::create() {
+      ResumeCall->setTailCallKind(CallInst::TCK_MustTail);
+    }
+
+    // Put a 'ret void' after the call, and split any remaining instructions to
```
if there are remaining instructions, this will silently elide them. I realize that there shouldn't be remaining instructions, but we could detect bugs like the PGO instrumentation + coro one (PR #71263) if that was asserted or WARNed or something more visible.

zmodem replied:
I'm hesitant to add such a check. I don't think it should be illegal to have instructions after the llvm.coro.await.suspend.handle call, the semantics is just that those instructions won't execute -- similarly to how instructions after other function calls may not run. For example one could imagine some kind of coverage instrumentation or debug intrinsics or whatnot; those would not run, meaning that code path wasn't taken.

For the PGO issue, if I understand correctly the issue was that it prevented the musttail lowering, which won't happen anymore.

mtrofin replied:
It's not about legality, it's about the fact they wouldn't be executed being a symptom of a bug - e.g. earlier pass (like coverage) believing there is something to cover while there isn't. We could at minimum have some "under build with asserts" behavior that (except for trivial cases like debug/lifetime instructions) says something.

zmodem replied:
I want the semantics to be well specified. Either the doc for this intrinsic should say that instructions between it and `llvm.coro.suspend` are not allowed (possibly with a list of exceptions) and we enforce it with an assert or the verifier; or it allows such instructions but says that they become unreachable as in the current version.

I'm in favor of the latter because it makes the IR more regular. We generally don't impose restrictions about what instructions are allowed between intrinsic calls. Instructions after a call not getting executed is not new, so I think what the current version of the patch does fits in well with the IR.

mtrofin replied:
Sorry for insisting on this, it's maybe because I got "bitten" before (with the suspend), but what other examples do we have where, silently, instructions don't get executed after a call?

Also, maybe this would become moot if we address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043 like @jyknight suggested (i.e. not even have the misleading edge)?

ChuanqiXu9 replied:
> Also, maybe this would become moot if we address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043 like @jyknight suggested (i.e. not even have the misleading edge)?

But IIRC, it is still possible that we'll have code inserted between `llvm.coro.await.suspend.{.*}` and `llvm.coro.suspend`, which is the problem we're discussing.

> Sorry for insisting on this, it's maybe because I got "bitten" before (with the suspend), but what other examples do we have where, silently, instructions don't get executed after a call?

If I read correctly, @zmodem said he'd like to mention this in the doc or check it by assertions or verifiers. So it looks consensus to me?

zmodem replied:
> Sorry for insisting on this

Not at all, I want us to resolve this, not paper over it. :)

>  it's maybe because I got "bitten" before (with the suspend)

But in that case the problem was kind of the opposite, right? Instructions between the `resume` and `suspend` blocked the musttail optimization. The problem was not whether those instructions would have been executed or not, but that they were inserted in a place where it wasn't allowed due to special constraints on this intrinsic.

The idea with my patch is to eliminate that problem by not having special restrictions about instructions after the intrinsic, they just won't be executed because control continues in the resumed function and doesn't come back.

> what other examples do we have where, silently, instructions don't get executed after a call?

This happens any time a call doesn't return.

> Also, maybe this would become moot if we address https://discourse.llvm.org/t/coro-pre-split-handling-of-the-suspend-edge/75043 like @jyknight suggested (i.e. not even have the misleading edge)?

Yes that's probably true, but that's a bigger change and I think my patch is a simpler solution to this specific problem: now that we have an intrinsic that corresponds directly to the symmetric transfer, let's just lower it as such directly.

> If I read correctly, @zmodem said he'd like to mention this in the doc or check it by assertions or verifiers. So it looks consensus to me?

No, I said that *if* we decide that instructions after `llvm.coro.await.suspend.handle` are not allowed, we should document and check it.

I still don't think we should do that, since it's a difficult invariant to maintain. We'd have to teach passes to respect it, keep updating a list of harmless instructions that are allowed (debug/lifetime intrinsics, ...) and we'd end up chasing down those assert/verifier failures forever.

With the current patch, we don't have to teach other passes anything. They already cannot assume that an unknown intrinsic call will return.

mtrofin replied:
> But in that case the problem was kind of the opposite, right? Instructions between the resume and suspend blocked the musttail optimization. The problem was not whether those instructions would have been executed or not, but that they were inserted in a place where it wasn't allowed due to special constraints on this intrinsic.
> The idea with my patch is to eliminate that problem by not having special restrictions about instructions after the intrinsic, they just won't be executed because control continues in the resumed function and doesn't come back.

I think I see where the disconnect comes from. You're referring to the desired outcome for coro: tailcalls being ensured. I agree with that, but I'm concerned with maintenance: suppose some other pass does what PGOInstrument did, in the future. Just like with coro, the potential for a undesired interaction between coro and their pass isn't known to them. How easy would it be to discover they made incorrect assumptions in their pass, diagnose and fix?

zmodem replied:
I'm also concerned with maintenance, that's why I'm pushing back against adding special rules around this intrinsic.

Let's say that we did add an assert checking for non-trivial instructions after `llvm.coro.await.suspend.handle`. If someone were to write a pass which inserts instrumentation after each call instruction to see if the call returns, they now need to treat coros as a special case. (Assuming they notice, by testing their pass on a coroutine.)

It's better if other passes don't have to have special cases for coroutines. I think that's the case with my current patch, because it fits within the existing IR model: function calls don't necessarily return, and when they don't, instructions after the call won't run and the optimizer may remove them. Here's a concrete example: https://godbolt.org/z/M65Ebcfh5

mtrofin replied:
In your example, if they derive some result based on the initial set of instrumentations they insert, they may produce reports that say "incomplete coverage" (for example). Some level of knowledge ends up being passed down to someone - maybe the user, maybe the analyzer of the data.

But I get the idea that maybe the domain is OK with loss of instructions, so the assert could be annoying. Dunno, only saw the example where it wasn't.

How about an ORE? The output could technically be usable by a tool in a fictitious coverage case, it offers more focused observability than dumping IR before/after in a debugging case, and it serves as self-documentation in the pass itself that "ya, we're dropping stuff here. It's intentional". The ORE wouldn't need to be too complex, maybe if we detect non-debug (and maybe also exclude lifetime) instructions, ORE out. WDYT?

zmodem replied:
An optimization remark would be less disruptive for sure, but I'm still not sure that it's right. Even though remarks are much lower level than regular diagnostics, they're still targeted at the compiler end user. If we emit a remark that says "deleted instructions after llvm.coro.await.suspend.handle call", what would the user do with that?

mtrofin replied:
For a fictitious example using some concrete actors: they could author a coverage analysis tool that would know to look for that and not misreport (say) 66% coverage because one of the edges was never traversed (because it never really existed)?

zmodem replied:
That still sounds like the remark would not target the compiler user, but the developer of the compiler tool -- and that's not really what optimization remarks are for.

The more I think about this, the less I like the idea of a remark. A "deleted instructions after llvm.coro.await.suspend.handle call" remark would be pure noise for a normal optimization remarks user. Plus we would need to maintain the code which to emit it, including tests for when it should trigger or not.

I think my patch makes the semantic of llvm.coro.await.suspend.handle clear in the doc, I think that behavior is compatible with how calls in the IR already works, and I think existing passes will handle it well. I don't think it needs special handling.

mtrofin replied:
> That still sounds like the remark would not target the compiler user, but the developer of the compiler tool -- and that's not really what optimization remarks are for.

Not sure if that's the case anymore, they do have a yaml outputter.

But that's besides the point. I think I can see this your way: in our running example, if I worked on coverage, and knew to write the ORE post-processor, I would also know to fix my coverage instrumenter to recognize not to place instructions in the wrong place. (I realize there are other cases where passes are OK to stay agnostic despite losing instructions here, but in the coverage scenario that's not the case any way you look at it). The main thing is how to figure I may have a problem, and really the fix there is "integration testing", really.

zmodem committed:
Merge remote-tracking branch 'origin/main' into coro_aggressive_musttail.squash2

zmodem committed:
remove clang/.../coro-symmetric-transfer-01.cpp

It was testing that "the lifetime of the coroutine handle used to obtain
the address is contained within single basic block, and hence does not
live across suspension points"

Since we no longer separately obtain the handle and pass it to
@llvm.coro.resume, this test doesn't really test anything.

zmodem committed:
tweaks

zmodem commented on clang/test/CodeGenCoroutines/coro-symmetric-transfer-01.cpp:
This was testing that "the lifetime of the coroutine handle used to obtain
the address is contained within single basic block, and hence does not
live across suspension points"

Since we no longer separately obtain the handle and pass it to
@llvm.coro.resume, this test doesn't really test anything.

zmodem commented on llvm/test/Transforms/Coroutines/coro-preserve-final.ll:
I wasn't able to figure out what this test does. There is a Clang level test which was added at the same time: llvm/test/Transforms/Coroutines/coro-preserve-final.ll presumable that's enough.

zmodem commented on llvm/test/Transforms/Coroutines/coro-split-musttail5.ll:
With my patch, some of these tests become less interesting. The class of problems arising from instructions between the resume and suspend doesn't exist with the new lowering. We could probably drop coro-split-musttail{5,6,7}.ll?

zmodem commented on llvm/test/Transforms/Coroutines/coro-split-musttail.ll:
```diff
@@ -40,10 +38,8 @@ exit:
 ; CHECK-LABEL: @f.resume(
-; CHECK: %[[addr2:.+]] = call ptr @llvm.coro.subfn.addr(ptr null, i8 0)
-; NOPGO-NEXT: musttail call fastcc void %[[addr2]](ptr null)
-; PGO: call void @llvm.instrprof
```
These `llvm.instrprof.value.profile` calls between the `llvm.coro.subfn.addr` call and tail call disappear with my patch. That's because the test runs `pgo-instr-gen` before `coro-split`. Is that representative of how the normal pipeline looks?

If the instrumentation is done after coro-split, it will see the indirect call and add the `llvm.instrprof.value.profile` as expected.

zmodem replied:
I looked into this, and I don't think my patch changes the situation for value profiling. PGO instrumentation is performed early. Before my patch it would see a call to `@llvm.coro.resume`, and now it will see `@llvm.coro.await.suspend.handle`. In neither case does it see the lowered indirect call to the resumed function, and so no `@llvm.instrprof.value.profile` call gets inserted. Supporting that would be a nice future improvement.

zmodem commented:
I've cleaned up the patch a bit and added some comments. Please take a look.

zmodem commented:
I think my patch is a significant improvement, both in terms of simplicity and reliability of the codegen for symmetric transfer, and would like to move forward. @ChuanqiXu9 @mtrofin do you have any further comments?

mtrofin commented:
> I think my patch is a significant improvement, both in terms of simplicity and reliability of the codegen for symmetric transfer, and would like to move forward. @ChuanqiXu9 @mtrofin do you have any further comments?

Nothing on my side, lgtm for my narrow concern. Please wait for @ChuanqiXu9 's lgtm tho.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -206,14 +210,37 @@ static void lowerAwaitSuspend(IRBuilder<> &Builder, CoroAwaitSuspendInst *CB) {
+      Intrinsic::coro_await_suspend_handle) {
+    // Follow the await_suspend by a lowered resume call to the returned
+    // coroutine.
+    if (auto *Invoke = dyn_cast<InvokeInst>(CB))
+      Builder.SetInsertPoint(Invoke->getNormalDest()->getFirstInsertionPt());
```
What if the CB is a call inst? It may be possible if the exception are disabled.

zmodem replied:
In that case the Builder's insert point is already in the right place.

That's what the "Follow ..." comment refers to. In both cases, we want to follow the await_suspend. It's only for an invoke that we need to jump to the next block.

ChuanqiXu9 replied:
Oh, got it. Maybe it might be better to make it more clear.

zmodem replied:
Yes, I will update the comment.

ChuanqiXu9 reviewed (approved):
LGTM with a comment.

zmodem committed:
tweak 'Follow the await_suspend' comment

zmodem committed:
Merge remote-tracking branch 'origin/main' into coro_aggressive_musttail.squash2

ChuanqiXu9 commented:
I received some reports for this patch breaking some codes. I am reproducing it. It looks like related to exceptions.

zmodem commented:
Thanks, let me know if I can help.

ChuanqiXu9 commented:
It is pretty interesting that I can pass the test by:

```
$git diff -U10
diff --git a/llvm/lib/Transforms/Coroutines/CoroSplit.cpp b/llvm/lib/Transforms/Coroutines/CoroSplit.cpp
index 450ea8234371..003bfbf7138a 100644
--- a/llvm/lib/Transforms/Coroutines/CoroSplit.cpp
+++ b/llvm/lib/Transforms/Coroutines/CoroSplit.cpp
@@ -220,20 +220,21 @@ static void lowerAwaitSuspend(IRBuilder<> &Builder, CoroAwaitSuspendInst *CB,
     }

     coro::LowererBase LB(*Wrapper->getParent());
     auto *ResumeAddr = LB.makeSubFnCall(NewCall, CoroSubFnInst::ResumeIndex,
                                         &*Builder.GetInsertPoint());

     LLVMContext &Ctx = Builder.getContext();
     FunctionType *ResumeTy = FunctionType::get(
         Type::getVoidTy(Ctx), PointerType::getUnqual(Ctx), false);
     auto *ResumeCall = Builder.CreateCall(ResumeTy, ResumeAddr, {NewCall});
+    ResumeCall->setCallingConv(CallingConv::Fast);

     // We can't insert the 'ret' instruction and adjust the cc until the
     // function has been split, so remember this for later.
     Shape.SymmetricTransfers.push_back(ResumeCall);

     NewCall = ResumeCall;
   }

   CB->replaceAllUsesWith(NewCall);
   CB->eraseFromParent();
```

The breaking only happens with optimizations in a workload.

ChuanqiXu9 commented:
Oh, it looks like some bots are not happy. I send the PR at https://github.com/llvm/llvm-project/pull/93167
